### DISCUSSION TITLE
Requesting Assistance with getting a private key needed to decrypt using rsa algorithm

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#23589] [Q&A] [answered]
Hi 
We previously  bsafe libs . We are migrating to openssl . Can you please suggest how to do the below operation using openssl 3.0

Previously we used below to get private key

``` C
B_KEY_OBJ GetPrivateKey()
{
	ITEM keyItem;
	unsigned char keyData[] = {key hex values};
    keyItem.len = 343;
    keyItem.data = (unsigned char *) T_malloc (keyItem.len);
    T_memcpy (keyItem.data, keyData, keyItem.len);

	B_KEY_OBJ key = (B_KEY_OBJ)NULL_PTR;
	if (0 != B_CreateKeyObject(&key))
		return NULL;

	if (0 != B_SetKeyInfo(key, KI_PKCS_RSAPrivateBER, (POINTER)&keyItem))
		return NULL;

	return key;
}
```

WIth OpenSSL 3.0 i am getting crash in below code, please suggest what is wrong with below code. If there are some resources which can help with migration of code

``` C
EVP_PKEY* get_pkey(OSSL_LIB_CTX* libctx, const char* propq)
{
	const unsigned char pData[] = {0xHex values};

	size_t data_len= sizeof(pData) / sizeof(pData[0]);
	EVP_PKEY* result = NULL;
	

	d2i_PrivateKey(EVP_PKEY_RSA, &result, (const unsigned char**)( & pData), data_len);

	return result;
}
```

**Comment by levitte** [comment, marked as answer]
The first thing I notice is this construction: `(const unsigned char**)( & pData)`

Despite your cast, this doesn't give you what you're trying to cast (as a matter of fact, I imagine that the cast is there because the compiler would [rightfully] protest otherwise).  See, since `pData` is an array, `pData` == `& pData`; taking the address of an array will still only get you the address of the array, not the address of a pointer to that array.

So, how about trying this:

``` C
EVP_PKEY* get_pkey(OSSL_LIB_CTX* libctx, const char* propq)
{
	const unsigned char pData[] = {0xHex values};
	const unsigned char *ppData = pData;

	size_t data_len= sizeof(pData) / sizeof(pData[0]);
	EVP_PKEY* result = NULL;
	

	d2i_PrivateKey(EVP_PKEY_RSA, &result, &ppData, data_len);

	return result;
}
```

**Reply by levitte** [reply]
Note: I haven't tested anything, and am only suggesting a path forward.  I leave trying it out to you.
